### ISSUE TITLE
Server-specific commands always visible

### ISSUE THREAD

**Issue by rouanth** [kak-lsp/kak-lsp#572]
Steps to reproduce:
* Open `kak`;
* Type `:ccls`;
* Observe the autocompletion containing commands that are only relevant for C/C++/etc. projects, and only with CCLS installed.

Same with `clangd`, `ejdtls` etc.

**Comment by sidkshatriya** [comment]
All kakoune commands are always visible. Ideally they would hide themselves depending on the context but they don't.

**Comment by rouanth** [comment]
Not true. Try launching `kak` in tmux and outside of it, for example.

**Comment by sidkshatriya** [comment]
@rouanth You are correct about tmux -- thanks for the counterexample. 

Let me modify my statement then: people _could_ write `define-command`s to be context sensitive i.e. only be visible in appropriate situations but generally I don't see it happen very much.

**Comment by sidkshatriya** [comment]
@rouanth You are correct about tmux -- thanks for the counterexample.

Also one more thing occurred to me. In the case of tmux being present or not, that is not a dynamic property. My guess is that when kakoune starts up, it checks for the presence of tmux and then `define-command` tmux specific commands (or not). 

Knowing when to show a specify command like `clangd-switch-source-header` is a more dynamic property. You should display this command only when the filetype of a buffer is C/C++ source/header perhaps. 

In other words, we just _could_ have `define-command -override -hidden clangd-switch-source-header ...` when you want to hide it and `-override` without the `-hidden` to unhide it. I wonder how practical that is or if there is a smarter way to do this. This is an interesting Kakoune problem to me!

**Comment by rouanth** [comment]
> My guess is that when kakoune starts up, it checks for the presence of tmux and then `define-command` tmux specific commands (or not).

This is more or less the case. More accurately, the commands are defined in a *module* (in `rc/windowing/tmux.kak`), and the module is imported only if tmux is detected (in `rc/windowing/detection.kak`). The same is performed for command available for specific filetypes (see `rc/tools/python/jedi.kak` for an example; you're interested in `require-module` and `provide-module` at the top).

> I wonder how practical that is or if there is a smarter way to do this.

Looks like the idiomatic way to do this is to define a module (say, `lsp-ccls`) with the commands that should be unavailable sometimes and only import it if the file is in the C family.

**Comment by sidkshatriya** [comment]
> Looks like the idiomatic way to do this is to define a module (say, lsp-ccls) with the commands that should be unavailable sometimes and only import it if the file is in the C family.

I'm not sure this approach would work in the way you we want. Because once imported, if you open another file which is _not_ a part of the C-family the command will still be available in the menu!

We would need to do the `define-command -override -hidden ...` trick to hide the command. And override again, _without_ `-hidden` when we want to show it...

**Comment by krobelus** [comment]
> > Looks like the idiomatic way to do this is to define a module (say, lsp-ccls) with the commands that should be unavailable sometimes and only import it if the file is in the C family.
> 		
> I'm not sure this approach would work in the way you we want. Because once imported, if you open another file which is _not_ a part of the C-family the command will still be available in the menu!

It's still good enough for many cases: it gets rid of most unnecessarily eager script evaluation, which I think is the main point of modules.

I have considered using a module for all of kak-lsp but it probably won't improve much, because we don't have heavy-weight initialization anyway, and users already load us explicitly, unlike some standard scripts.

A lsp-ccls module sounds promising, but we'd need a hook that loads the module when we first open a C/C++ file and ccls is installed.
I guess we could just not load the commands if ccls is not in $PATH.
So far we don't have shell calls at startup but one or two won't hurt.

Of course we could make everything a subcommand of a new `lsp-ccls` command that loads everything, but that command is still ugly.
Not really happy with any of the solutions.

For now a workaround could be

```kak
evaluate-commands %sh{
    kak-lsp --kakoune -s $kak_session |
        sed '/define-command.*ccls\|clangd\|ejdtls/s/^/nop /'
}
```
